gazebo/Server.cc

```diff
@@ -542,6 +548,110 @@ void Server::Fini()
   gazebo::shutdown();
 }
 
+std::map<std::string, gazebo::common::Time> sensorsLastMeasurementTime;
+std::map<std::string, gazebo::common::Time> worldLastMeasurementTime;
+struct sensorPerformanceMetricsType
+{
+  double sensorRealUpdateRate;
+  double sensorSimUpdateRate;
+  double sensorFPS;
+};
+std::map<std::string, struct sensorPerformanceMetricsType> sensorPerformanceMetrics;
+/// \brief Publisher for run-time simulation performance metrics..
+transport::PublisherPtr performanceMetricsPub;
+
+void PublishPerformanceMetrics()
+{
+  physics::WorldPtr world = physics::get_world("default");
+
+  /// Outgoing run-time simulation performance metrics.
+  msgs::PerformanceMetrics performanceMetricsMsg;
+
+  // Real time factor
+  common::Time realTime = world->RealTime();
+  common::Time simTime = world->SimTime();
+  if (realTime == 0)
+    simTime = 0;
+  else
+    simTime = simTime / realTime;
+
+  if (simTime > 0)
+    performanceMetricsMsg.set_real_time_factor(simTime.Double());
+  else
+    performanceMetricsMsg.set_real_time_factor(0.0);
+
+  /// update sim time for sensors
+  for (auto model: world->Models())
+  {
+    for (auto link: model->GetLinks())
+    {
+      for (unsigned int i = 0; i < link->GetSensorCount(); i++)
+      {
+        std::string name = link->GetSensorName(i);
+        sensors::SensorPtr sensor = sensors::get_sensor(name);
+
+        auto ret = sensorsLastMeasurementTime.insert(
+            std::pair<std::string, gazebo::common::Time>(name, 0));
+        worldLastMeasurementTime.insert(
+                std::pair<std::string, gazebo::common::Time>(name, 0));
+        if (ret.second==false)
+        {
+          double updateSimRate =
+            (sensor->LastMeasurementTime() - sensorsLastMeasurementTime[name]).Double();
+          double updateRealRate =
+            (world->RealTime() - worldLastMeasurementTime[name]).Double();
+          sensorsLastMeasurementTime[name] = sensor->LastMeasurementTime();
+          if (updateSimRate > 0.0)
+          {
+            struct sensorPerformanceMetricsType emptySensorPerfomanceMetrics;
+            auto ret2 = sensorPerformanceMetrics.insert(
+                std::pair<std::string, struct sensorPerformanceMetricsType>
+                  (name, emptySensorPerfomanceMetrics));
+            if (ret2.second==false)
+            {
+              sensorPerformanceMetrics[name].sensorSimUpdateRate = 1.0/updateSimRate;
```

Comment:
I change the physics update rate to 100 instead of 1000 (from gui client left panel: `Physics` -> `real time update rate`), I expect the sensor sim update rate to be the same as what's specified. But with a camera sensor, I see that the sim update rate  is a small value close to 0 and the real update rate is larger than expected. 

It's working correctly for non-rendering sensors though.

Reply:
I thing that there are two cases:
 - Without lockstep: I think the issue that you are explaining could happen when you are running Gazebo without the lockstep flag. There is no garanties when the callback is called.
 - lockstep: I change the physics update rate to 100 the values remain stable

@iche033  Does it make sense?

Reply:
yeah that makes sense. It could be confusing to users that the metrics are correct for some sensors (physics based sensors) but not others (rendering sensors) since they may not be familiar with the internals of gazebo. We may have to have some documentation to explain that the metrics for rendering senors like cameras are more accurate in lock step mode.

Reply:
I also just noticed that when I change the real time update rate from 1000 to 100, the `real_tme_factor` values in the published metrics msg takes a while (minutes) to get to the correct factor of 0.1. I think that's because it's computed based on total sim time / real time - so it's giving the overall RTF for the entire simulation. The instantaneous value maybe more useful here.

Reply:
Added a tutorial in https://github.com/osrf/gazebo_tutorials/pull/117